weblab/datasets/views.py

```diff
@@ -370,3 +372,110 @@ def get_context_data(self, **kwargs):
         ]
 
         return super().get_context_data(**kwargs)
+
+
+class DatasetMapColumnsView(UserPassesTestMixin, VisibilityMixin, DetailView):
+    model = Dataset
+    template_name = 'datasets/map_columns.html'
+
+    # Raise a 403 error rather than redirecting to login,
+    # if the user isn't allowed to modify dataset
+    raise_exception = True
+
+    def test_func(self):
+        return self.get_object().is_editable_by(self.request.user)
+
+    def _get_object(self):
+        if not hasattr(self, 'object'):
+            self.object = self.get_object()
+        return self.object
+
+    def get_forms(self):
+        """
+        Construct the forms for this view.
+
+        Populates `self.forms` which is a dict mapping protocol version
+        to a list of forms for that protocol version. Each protocol version
+        has one form per dataset column.
+
+        @return self.forms
+        """
+        dataset = self._get_object()
+
+        self.forms = {}
+
+        # Group existing mappings by protocol version
+        existing_mappings = dataset.column_mappings.all()
+        mappings_by_version = {
+            version: list(mappings)
+            for version, mappings in groupby(existing_mappings, lambda x: x.protocol_version)
+        }
+
+        # Full list of protocol versions for this user
+        protocol_versions = dataset.protocol.cachedentity.versions.visible_to_user(self.request.user)
+
+        for version in protocol_versions:
+            self.forms[version] = []
+
+            # Each form will ist all but FLAG ioputs for this protocol version
+            ioputs = version.ioputs.exclude(kind=(ProtocolIoputs.FLAG))
+
+            # Group existing mappings for this version by column name
```

Comment:
'Group' might be misleading, since the unique_together means we'll never have multiples?